warehouse/router/sync_test.go

```diff
@@ -0,0 +1,261 @@
+	if m.err != nil {
+		return model.Schema{}, m.err
+	}
+	return m.schemaInWarehouse, nil
+}
+
+type mockSyncSchemaRepo struct {
+	getLocalSchemaerr    error
+	updateLocalSchemaErr error
+	fetchSchemaErr       error
+	hasChanged           bool
+	schemaMap            map[string]model.Schema
+}
+
+func schemaKey(sourceID, destinationID, namespace string) string {
+	return fmt.Sprintf("%s_%s_%s", sourceID, destinationID, namespace)
+}
+
+func (m *mockSyncSchemaRepo) GetLocalSchema(_ context.Context) (model.Schema, error) {
+	if m.getLocalSchemaerr != nil {
+		return model.Schema{}, m.getLocalSchemaerr
+	}
+
+	return model.Schema{}, nil
+}
+
+func (m *mockSyncSchemaRepo) UpdateLocalSchemaWithWarehouse(_ context.Context, _ model.Schema) error {
+	if m.updateLocalSchemaErr != nil && m.hasChanged {
+		return m.updateLocalSchemaErr
+	}
+	return nil
+}
+
+func (m *mockSyncSchemaRepo) HasSchemaChanged(_ model.Schema) bool {
+	return m.hasChanged
+}
+
+func (m *mockSyncSchemaRepo) FetchSchemaFromWarehouse(_ context.Context, _ schema.FetchSchemaRepo) (model.Schema, error) {
+	if m.fetchSchemaErr != nil {
+		return nil, m.fetchSchemaErr
+	}
+
+	return m.schemaMap[schemaKey("test-sourceID", "test-destinationID", "test-namespace")], nil
+}
+
+func TestSync_SyncRemoteSchema(t *testing.T) {
+	t.Run("fetching schema from local fails", func(t *testing.T) {
+		mock := &mockSyncSchemaRepo{
+			getLocalSchemaerr: fmt.Errorf("error fetching schema from local"),
+		}
+		mockFetchSchema := &mockFetchSchemaRepo{}
+		r := &Router{
+			logger: logger.NOP,
+		}
+		err := r.syncRemoteSchema(context.Background(), mockFetchSchema, mock)
+		require.Error(t, err)
+		require.True(t, errors.Is(err, mock.getLocalSchemaerr))
+	})
+	t.Run("fetching schema from warehouse fails", func(t *testing.T) {
+		mock := &mockSyncSchemaRepo{
+			fetchSchemaErr: fmt.Errorf("error fetching schema from warehouse"),
+		}
+		mockFetchSchema := &mockFetchSchemaRepo{}
+		r := &Router{
+			logger: logger.NOP,
+		}
+		err := r.syncRemoteSchema(context.Background(), mockFetchSchema, mock)
+		require.Error(t, err)
+		require.True(t, errors.Is(err, mock.fetchSchemaErr))
+	})
+	t.Run("schema has changed and updating errors", func(t *testing.T) {
+		mock := &mockSyncSchemaRepo{
+			hasChanged:           true,
+			updateLocalSchemaErr: fmt.Errorf("error updating local schema"),
+		}
+		mockFetchSchema := &mockFetchSchemaRepo{}
+		r := &Router{
+			logger: logger.NOP,
+		}
+		err := r.syncRemoteSchema(context.Background(), mockFetchSchema, mock)
+		require.Error(t, err)
+	})
+	t.Run("schema has changed and updating errors", func(t *testing.T) {
+		mock := &mockSyncSchemaRepo{
+			hasChanged:           false,
+			updateLocalSchemaErr: fmt.Errorf("error updating local schema"),
+		}
+		mockFetchSchema := &mockFetchSchemaRepo{}
+		r := &Router{
+			logger: logger.NOP,
+		}
+		err := r.syncRemoteSchema(context.Background(), mockFetchSchema, mock)
+		require.NoError(t, err)
+	})
```

Comment:
w.r.t to the mock, didin't get the test case? UpdateLocalSchemaWithWarehouse will not return an error?
```GO
func (m *mockSyncSchemaRepo) UpdateLocalSchemaWithWarehouse(context.Context, model.Schema) error {
	if m.updateLocalSchemaErr != nil && m.hasChanged {
		return m.updateLocalSchemaErr
	}
	return nil
}
```